Allow for Kubernetes cluster authentication via google auth

RochesterinNYC opened the pull request:
- For the kubernetes and kubernetes-backend plugins
  - The kubernetes (front-end) plugin uses the googleAuthApi and goes
    through oauth flow to fetch a Google auth token for the current user
    and pass that in a request to the kubernetes-backend plugin, which
    uses this token as the service account token (and hence as the
    authentication token for making K8s API requests).
- Related to https://github.com/spotify/backstage/issues/2552

RochesterinNYC commented:
@mclarke47 for review as well.

RochesterinNYC commented:
Here's screenshots from running the front-end and backend for Backstage locally with these changes.

(This was using Matt Clarke's example catalog component at https://github.com/mclarke47/dice-roller/blob/master/catalog-info.yaml and his instructions for the Kubernetes plugin at https://github.com/spotify/backstage/tree/d66c6a7/plugins/kubernetes-backend/examples/dice-roller).

`dice-roller` Deployment/Pods running on a GKE cluster (Google auth used for auth when fetching this info via the K8s API):

<img width="1527" alt="Screen Shot 2020-10-01 at 2 33 44 AM" src="https://user-images.githubusercontent.com/1504464/94776628-10628c80-0390-11eb-88ab-c5a045aa4455.png">

`dice-roller` Deployment/Pods running on minikube (service account token used for auth when fetching this info via the K8s API):

<img width="1519" alt="Screen Shot 2020-10-01 at 2 33 52 AM" src="https://user-images.githubusercontent.com/1504464/94776800-620b1700-0390-11eb-9f75-b987e5661183.png">

Here's the additional `app-config.development.yaml` config I added for this local running:

```
kubernetes:
  clusterLocatorMethod: 'configMultiTenant'
  clusters:
    - url: https://<GKE cluster master IP>
      name: production-us-central1
      authProvider: google
    - url: https://192.168.64.2:8443
      name: minikube
      authProvider: serviceAccount
      serviceAccountToken: <service account token value>

auth:
  providers:
    google:
      development:
        clientId:
          $secret:
            env: <Google oauth app ID>
        clientSecret:
          $secret:
            env: <Google oauth app secret>
```

RochesterinNYC commented:
Usually, I'm a proponent of comprehensive PRs with multiple commits and wholly complete features, but in this case, I would like to get this PR out there for review as it's my first Backstage PR (for the open source version) and I'm not super well versed in typescript.

Some follow-up PRs/TODOs that I'll be making before https://github.com/spotify/backstage/issues/2552 can be considered done:

- Add validation for the `kubernetes.clusters` part of the config that runs at app start time that will:
  - Validate all clusters have an `authProvider` specified
  - Validate all clusters with `serviceAccount` as the `authProvider` have a `serviceAccountToken` value specified as well
- Provide a kubernetes-backend endpoint that the frontend can query to determine what kinds of auth providers from which auth tokens are required/what oauth flows need to be done to get (right now, regardless of what kinds of clusters are specified in the config, Google oauth flow will occur and a Google auth token will be fetched and passed to the backend)
- Docs for how to specify google auth in general and how to specify clusters to use google auth for API requests

Additionally, something I noticed that I'm confused about:
- Regardless of whether I am already generally logged in with Google to Backstage, clicking the Kubernetes tab for a component for the first time requires me to oauth again with Google. This might be due to how I'm invoking the google auth package or internal library. Is there a way to get a token or some cached token from that initial Google oauth/sign-in ?

mclarke47 commented on plugins/kubernetes-backend/src/cluster-locator/MultiTenantConfigClusterLocator.ts:
```diff
@@ -28,20 +29,45 @@ export class MultiTenantConfigClusterLocator
+      throw new Error(
+        `Unsupported Kubernetes auth provider "${authProviderKind}"`,
+      );
+    });
```
I'm a bit concerned that we're doing too much here. Possibly this method should return the clusterDetails without tokens (except those from the config) and we combine them in a separate method.

This will probably be necessary anyway when other `ClusterLocators` are brought in (for example the HTTP get one).

Rugvip replied:
Yeah definitely need to invert the responsibilities a bit here.

We also need to make sure it's possible to provide custom auth providers for organizations with their own setup or unique way to auth towards different clusters. Ideally it's a TypeScript interface that we provide a couple of implementations of in this project, but allow custom implementations to be passed in too.

mclarke47 commented on plugins/kubernetes/src/components/KubernetesContent/KubernetesContent.tsx:
```diff
@@ -105,16 +107,24 @@ export const KubernetesContent = ({ entity }: KubernetesContentProps) => {
+      const googleAuthToken: string = await googleAuth.getAccessToken(
+        'https://www.googleapis.com/auth/cloud-platform',
+      );
+      const authTokens: AuthTokens = { google: googleAuthToken };
```
We'll need error handling at some point here, and a place to call other auth providers.

mclarke47 commented:
1 or 2 comments but apart from that it's a great start! I'm really happy we have been able to use google auth tokens to talk to the GKE clusters 🎊

Rugvip commented on plugins/kubernetes-backend/src/cluster-locator/MultiTenantConfigClusterLocator.test.ts:
```diff
@@ -16,6 +16,7 @@
 
 import '@backstage/backend-common';
 import { MultiTenantConfigClusterLocator } from './MultiTenantConfigClusterLocator';
+import { AuthTokens } from '..';
```
Best make this more specific to avoid import cycles, as in avoid importing from a direct ancestor

RochesterinNYC replied:
Could you clarify what you mean here/what's an example of a more specific import?

Rugvip commented on plugins/kubernetes-backend/src/cluster-locator/MultiTenantConfigClusterLocator.ts:
```diff
@@ -28,20 +29,45 @@ export class MultiTenantConfigClusterLocator
           url: c.getString('url'),
           serviceAccountToken: c.getOptionalString('serviceAccountToken'),
+          authProvider: (c.getOptionalString('authProvider')
+            ? c.getOptionalString('authProvider')
```
Just `c.getOptionalString('authProvider') ?? 'serviceAccount'` here. I'd also add validation that it's actually a `AuthProviderType` at runtime instead of casting too

Rugvip commented on plugins/kubernetes-backend/src/service/router.ts:
```diff
@@ -62,15 +62,16 @@ export const makeRouter = (
+  router.post('/services/:serviceId', async (req, res) => {
     const serviceId = req.params.serviceId;
-
+    const authTokens: AuthTokens = req.body.auth;
```
We likely need to pass on the entire request object to whatever part of the backend is responsible for authenticating requests. The mapping would probably be something like `(req, service, cluster) -> credentials` or something of the sort. Not sure exactly where that would happen though.

Still happy to merge this PR and see how we can untangle things afterwards

Rugvip commented on plugins/kubernetes-backend/src/types/types.ts:
```diff
@@ -113,7 +117,10 @@ export interface KubernetesFetcher {
-  getClusterByServiceId(serviceId: string): Promise<ClusterDetails[]>;
+  getClusterByServiceId(
+    serviceId: string,
+    authTokens: AuthTokens,
```
As mentioned ample times already, this is pretty leaky xP

Rugvip commented on plugins/kubernetes/src/components/KubernetesContent/KubernetesContent.tsx:
```diff
@@ -105,16 +107,24 @@ export const KubernetesContent = ({ entity }: KubernetesContentProps) => {
-      });
-  }, [entity.metadata.name, kubernetesApi]);
+    (async () => {
+      // TODO: Query Kubernetes backend to get info on which auth providers are in use
```
We may want to provide a frontend interface for k8s auth too, as it would become a pretty clean place to switch between different methods in an app. We'd likely need something of the sort anyway to be able to support custom auth implementations.

What I really wanna figure out is if there's some good common contracts that we can put in place both between the frontend and backend, as well as internally within the backend too

Rugvip reviewed (commented):
🎉 , lots of comments, and should fix the import path thing, but otherwise I'm happy to merge and iterate

RochesterinNYC commented:
I thought a fair bit more about the abstraction/interface aspect of the feedback and what y'all were saying and put some brainstorming/a rough proposal for what the interfaces and contract could be between backend and frontend for this K8s auth stuff: https://github.com/spotify/backstage/issues/2552#issuecomment-702545382

Posted in that issue for a bit greater visibility/that issue seems like a better place for higher-level abstraction/interface discussion.

RochesterinNYC commented:
I added a commit that implements what I was proposing regarding KubernetesAuthProvider and KubernetesAuthTranslator interfaces and how they work at https://github.com/spotify/backstage/issues/2552#issuecomment-702545382. For the time being, I've implemented these interfaces and implementations of them for `serviceAccount` and `google` auth provider types, which works as of now in practice. I'd roughly suggest we go with that/try to get that merged and then iterate/change things around until we get to an interface that we're all happy with + is easy for Backstage developers/maintainers to use and extend?

RochesterinNYC commented on plugins/kubernetes/src/components/KubernetesContent/KubernetesContent.tsx:
```diff
@@ -105,15 +113,63 @@ export const KubernetesContent = ({ entity }: KubernetesContentProps) => {
+  for (const authProviderStr of authProviders) {
+    let api;
+    if (authProviderApiMapping.has(authProviderStr)) {
+      api = useApi(authProviderApiMapping.get(authProviderStr)!);
```
This causes this linting error to be thrown: `React Hook "useApi" may be executed more than once. Possibly because it is called in a loop. React Hooks must be called in the exact same order in every component render  react-hooks/rules-of-hooks`. I'm not sure how to deal with this in a way that doesn't involve loading in the same set of providers/APIs (ex. all the ones that we support for Kubernetes auth) each time/in a hardcoded manner.

It seems that any dynamic usage of useApi (ex. in a way that only involves loading in the K8s auth related APIs that are defined in the app-config) would result in this error being thrown/noted, due to inconsistent useApi invocation order.

freben replied:
Yeah that's one of the quirks of hooks in React in general because of how they are implemented - they have to always be invoked in the same order. That's why you'll almost always find them at the very top of component code, and never in a loop or conditional.

freben replied:
So I haven't looked at the rest of the PR yet, but it seems that you could store the provider key-to-instance map in a provider of sorts, that you can fetch once at the top as `useKubernetesAuthProviders()`.

If that sounds like circular reasoning (how would that one then be implemented), you could likely instantiate it at plugin boot time, something along the lines of

```ts
export const plugin = createPlugin({
  id: 'myKubernetesPlugin',
  apis: [
    createApiFactory({
      api: kubernetesAuthProvidersRef,
      deps: { configApi: configApiRef, googleApi: googleApiRef },
      factory({ configApi, googleApi }) {
        // ... create the various supported providers
        return new KubernetesAuthProviders(...);
      },
    }),
  ],
});
```

RochesterinNYC replied:
Gotcha, instantiation at plugin creation time makes sense. Would I be able to, in this section of logic (presumably in here: https://github.com/spotify/backstage/blob/2d73c037a0affd3b1c2eba6878543a37cd53ba1f/plugins/kubernetes/src/plugin.ts) load in the configApi outside of this `createPlugin` invocation, use it to read the `app-config` and get the list of different `authProvider`s that are defined across the various `kubernetes.clusters` in the config, and then create an `kubernetesAuthProvidersRef` that uses the various auth APIs correlated with the defined `authProvider`s ? 

Seems like I would be able to do this as this doesn't use `useApi`? And I would simply be able to do one `useApi(kubernetesAuthProvidersRef)` in the `KubernetesContent.tsx`. I'll give this a try later.

Rugvip replied:
@RochesterinNYC Did actually run in to this with the auth providers and thought about some kinda meta-factory that can read config and generate factories. Didn't go down that path though, because I don't see any big win with it and it adds a lot of complexity.

I think the sane way to go for now is to list all the auth APIs that you have implementations for, and then allow more granular access and customization by overriding the default factory in the app.

A possible general solution I think could be the ability to optionally depend on APIs, as in if they're available in the app, great, use it, but if it isn't , we don't care. That or lazy loading, or both. That's something we could look at bringing in if we feel there's a need, both for the `useApi` and `createApiFactory` routes.

Either way you need to enable the possibility for the consumer to provide their own APIs in some way, because an organization that uses Backstage may have their completely custom internal authentication provider, implemented with a custom and unknown to us `ApiRef`.

RochesterinNYC replied:
I amended the top commit to basically create a `KubernetesAuthProviders` "API" as @freben suggested that stores/holds instances of the `KubernetesAuthProvider`s that are currently supported and is created at plugin time instead of `KubernetesComponent` render/creation time.

The google auth workflow still works as expected.

Ready for review again @mclarke47 @Rugvip. I think we should at some point seek to merge this in as it's currently working for Google auth and then make more PRs/improvements and etc. instead of aiming to make this one massive, all-encompassing PR that fits exactly how we want to do Kubernetes auth abstractions/interface/UX wise overall.

mclarke47 replied:
Yeah I'm up for shipping and iterating too

RochesterinNYC commented on plugins/kubernetes/src/components/KubernetesContent/KubernetesContent.tsx:
```diff
@@ -105,15 +113,63 @@ export const KubernetesContent = ({ entity }: KubernetesContentProps) => {
+        .catch(e => {
+          setError(e.message);
+        });
+    })();
```
It seems I need to add `authProviders` and `kubernetesAuthProviderMap` to this dependency list for `useEffect`. However, doing so seems to cause the Kubernetes page to re-render the KubernetesContent component (and hence make Kubernetes resource fetch requests in an endless loop). Via `console.log` usage that logged all 4 of these dependencies with each render/invocation of the `useEffect` hook, it didn't seem like there were any differences in the values of any of these 4 variables, so I'm not sure why the re-rendering is occurring.

Will need to look more into this possibly as the infinite looping is inefficient/a strain on the backend.

RochesterinNYC replied:
Added into the comment a comment to have the linter ignore the need to have `authProviders` in the dependency array (I saw the `eslint-disable react-hooks/exhaustive-deps` being used in a few other plugins).

mclarke47 commented:
:shipit: && 🇮🇹 👂 🎱

Rugvip commented on plugins/kubernetes/src/kubernetes-auth-provider/types.ts:
```diff
@@ -0,0 +1,38 @@
+  ): Promise<AuthRequestBody>;
+}
+
+export const KubernetesAuthProvidersApiRef = createApiRef<
```
```suggestion
export const kubernetesAuthProvidersApiRef = createApiRef<
```

RochesterinNYC replied:
Updated the latest commit with this change!

Rugvip reviewed (approved):
Ship & It :grin:

Minor nit but could fix later too

RochesterinNYC commented:
I don't have write access to this repo so I can't merge, can someone with proper access merge please?
